zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+                       committedCounter, tempVarsZ, followerVarsZ, verifyVarsZ, electionMsgs, idTable>>
+
+FollowerHandleUPTODATE(i, j) ==
+        /\ state[i] = FOLLOWING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = UPTODATE
+        /\ LET msg == msgs[j][i][1]
+               infoOk == /\ leaderAddr[i] = j
+                         /\ acceptedEpoch[i] = msg.mepoch
+               correct == /\ infoOk
+                          /\ zabState[i] = SYNCHRONIZATION
+                          /\ currentEpoch[i] = msg.mepoch
+           IN /\ infoOk
+              /\ \/ /\ correct
+                    /\ commitIndex' = [commitIndex EXCEPT ![i] = Maximum({commitIndex[i], msg.mcommit})] \* May have received COMMIT and have a bigger commit index.
+                    /\ zabState'    = [zabState    EXCEPT ![i] = BROADCAST]
+                    /\ UNCHANGED inherentViolated
+                 \/ /\ ~correct
+                    /\ PrintT("Exception: Condition correct is false in FollowerHandleUPTODATE(" \o ToString(i) \o ", " \o ToString(j) \o ").")
+                    /\ inherentViolated' = TRUE
+                    /\ UNCHANGED <<commitIndex, zabState>>
+        /\ Discard(j, i)
+        /\ UNCHANGED <<state, currentEpoch, lastZxid, acceptedEpoch, history, electionVarsZ, leaderVarsZ, tempVarsZ, followerVarsZ,
+                       proposalMsgsLog, epochLeader, electionMsgs, idTable>>
+
+-----------------------------------------------------------------------------
+(* In phase l31, leader receives client request and broadcasts PROPOSAL.
+   Note: In production, any server in traffic can receive requests and forward it to leader if necessary.
+   We choose to let leader be the sole one who can receive requests, to simplify spec and keep correctness at the same time. *)
+ClientRequest(i, v) ==
+        /\ state[i] = LEADING
+        /\ zabState[i] = BROADCAST
+        /\ currentCounter' = [currentCounter EXCEPT ![i] = currentCounter[i] + 1]
+        /\ LET newTransaction == [epoch   |-> acceptedEpoch[i],
+                                  counter |-> currentCounter'[i],
+                                  value   |-> v]
+           IN /\ history'  = [history  EXCEPT ![i] = Append(history[i], newTransaction)]
+              /\ lastZxid' = [lastZxid EXCEPT ![i] = <<acceptedEpoch[i], currentCounter'[i]>>]
+              /\ ackIndex' = [ackIndex EXCEPT ![i][i] = Len(history'[i])]
+              /\ UpdateProposal(i, i, lastZxid'[i], currentEpoch[i])
+        /\ UNCHANGED <<state, currentEpoch, zabState, acceptedEpoch, commitIndex, logicalClock, receiveVotes, outOfElection, recvQueue, waitNotmsg,
+                       leadingVoteSet, learners, cepochRecv, ackeRecv, ackldRecv, forwarding, sendCounter, committedIndex, committedCounter,
+                       tempVarsZ, followerVarsZ, verifyVarsZ, msgVarsZ, idTable>>
+
+LeaderBroadcast1(i) ==
```

Comment:
Action name is not descriptive, perhaps rename to something like LeaderBroadcastProposal